GraphQL Batch processing throws a ContextNotActiveException

**Describe the bug**
GraphQL Batch processing throws a ContextNotActiveException

**Expected behavior**
As described in https://quarkus.io/guides/microprofile-graphql#batching Batch processing should work =)

**Actual behavior**
GraphQL Batch processing throws a ContextNotActiveException

**To Reproduce**
See attachment for simple setup:
[TestGraphQLResource.zip](https://github.com/quarkusio/quarkus/files/5567840/TestGraphQLResource.zip)

Simple Source relations like the following work as expected:
```
    public Yyy getYyy(@Source Xxx xxx) {
        return mapYyy(xxx);
    }

    public List<Zzz> getZzz(@Source Xxx xxx) {
        return mapZzz(xxx);
    }
```

When changing to Batch processing a ContextNotActiveException is thrown.
```
    public List<Yyy> getYyy(@Source List<Xxx> xxxList) {
        return xxxList.stream().map(this::mapYyy).collect(Collectors.toList());
    }

    public List<List<Zzz>> getZzz(@Source List<Xxx> xxxList) {
        return xxxList.stream().map(this::mapZzz).collect(Collectors.toList());
    }
```


**Configuration**
No properties set for GraphQL

**Screenshots**
```
2020-11-19 11:29:59,847 ERROR [io.sma.graphql] (ForkJoinPool.commonPool-worker-13) SRGQL012000: Data Fetching Error: java.util.concurrent.CompletionException: java.lang.RuntimeException: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.RuntimeException: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader.doSourceCall(SourceBatchLoader.java:83)
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader.access$000(SourceBatchLoader.java:27)
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader$1.run(SourceBatchLoader.java:63)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader.invokePrivileged(SourceBatchLoader.java:56)
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader.lambda$load$0(SourceBatchLoader.java:49)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        ... 6 more
Caused by: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
        at whatever.wherever.TestGraphQLResource_ClientProxy.arc$delegate(TestGraphQLResource_ClientProxy.zig:68)
        at whatever.wherever.TestGraphQLResource_ClientProxy.getZzz(TestGraphQLResource_ClientProxy.zig:249)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.smallrye.graphql.execution.datafetcher.helper.ReflectionHelper.invoke(ReflectionHelper.java:45)
        at io.smallrye.graphql.execution.batchloader.SourceBatchLoader.doSourceCall(SourceBatchLoader.java:80)
        ... 12 more
```

**Environment (please complete the following information):**

- Output of uname -a or ver:
- Output of java -version:
   openjdk version "11.0.7" 2020-04-14
   OpenJDK Runtime Environment GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02) 
   OpenJDK 64-Bit Server VM GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02, mixed mode, sharing)
- GraalVM version (if different from Java):
- Quarkus version or git rev: 1.9.2
- Build tool (ie. output of mvnw --version or gradlew --version): 3.6.3

**Additional context**
(Add any other context about the problem here.)

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -42,7 +42,7 @@
         <smallrye-health.version>2.2.5</smallrye-health.version>
         <smallrye-metrics.version>2.4.4</smallrye-metrics.version>
         <smallrye-open-api.version>2.0.16</smallrye-open-api.version>
-        <smallrye-graphql.version>1.0.18</smallrye-graphql.version>
+        <smallrye-graphql.version>1.0.19</smallrye-graphql.version>
         <smallrye-opentracing.version>1.3.4</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>4.3.2</smallrye-fault-tolerance.version>
         <smallrye-jwt.version>2.4.0</smallrye-jwt.version>

--- a/extensions/smallrye-graphql/deployment/pom.xml
+++ b/extensions/smallrye-graphql/deployment/pom.xml
@@ -48,6 +48,11 @@
             <groupId>io.smallrye</groupId>
             <artifactId>smallrye-graphql-ui-graphiql</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-context-propagation-deployment</artifactId>
+        </dependency>
         
         <!--  Test dependencies -->
         <dependency>

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLCDIContextPropagationTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLCDIContextPropagationTest.java
@@ -1,0 +1,188 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.context.ManagedExecutor;
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Name;
+import org.eclipse.microprofile.graphql.Query;
+import org.eclipse.microprofile.graphql.Source;
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
+
+/**
+ * Testing scenarios which require CDI context propagation to work under the hood.
+ */
+public class GraphQLCDIContextPropagationTest extends AbstractGraphQLTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(TestPojo.class, ResourceThatNeedsCdiContextPropagation.class)
+                    .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
+
+    /**
+     * Call a query which needs the CDI context to be propagated, because it is on a RequestScoped bean
+     * and involves retrieving a batch source field (these are retrieved asynchronously).
+     */
+    @Test
+    public void testCdiContextPropagationForBatchSources() {
+        String pingRequest = getPayload("{\n" +
+                "  pojos {\n" +
+                "    duplicatedMessage\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(pingRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.pojos.duplicatedMessage", Matchers.contains("AA", "BB"));
+    }
+
+    /**
+     * Same as above, but the batch source returns a list of CompletionStages.
+     */
+    @Test
+    public void testCdiContextPropagationForBatchSourcesAsync() {
+        String pingRequest = getPayload("{\n" +
+                "  pojos {\n" +
+                "    duplicatedMessageAsync\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(pingRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.pojos.duplicatedMessageAsync", Matchers.contains("AA", "BB"));
+    }
+
+    /**
+     * In this case, there is an asynchronous query and the application launches a CompletableFuture,
+     * therefore the application is responsible for propagating the context to that future.
+     */
+    @Test
+    public void testManualContextPropagation() {
+        String pingRequest = getPayload("{\n" +
+                "  pojo_manual_propagation {\n" +
+                "    message\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(pingRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.pojo_manual_propagation.message", Matchers.equalTo("A"));
+    }
+
+    @GraphQLApi
+    @ApplicationScoped
+    public static class ResourceThatNeedsCdiContextPropagation {
+
+        /**
+         * This is to make sure that `getPojos` and `duplicatedMessage` are called on the same class instance - they
+         * need to share the same CDI request context rather than create a new one for calling `duplicatedMessage`.
+         */
+        private volatile Long injectedBeanId;
+
+        @Inject
+        InjectedBean injectedBean;
+
+        @Inject
+        ManagedExecutor executor;
+
+        @Query("pojos")
+        public List<TestPojo> getPojos() {
+            this.injectedBeanId = injectedBean.getId();
+            List<TestPojo> pojos = new ArrayList<>();
+            pojos.add(new TestPojo("A"));
+            pojos.add(new TestPojo("B"));
+            return pojos;
+        }
+
+        @Query("pojo_manual_propagation")
+        public CompletionStage<TestPojo> getPojoWithProgrammaticContextPropagation() {
+            Long id = injectedBean.getId();
+            return executor.supplyAsync(
+                    () -> {
+                        if (!injectedBean.getId().equals(id)) {
+                            throw new IllegalStateException("The future was not executed in the correct request context");
+                        }
+                        return new TestPojo("A");
+                    });
+        }
+
+        /**
+         * This source field duplicates the message of a TestPojo (repeats it twice).
+         */
+        @Name("duplicatedMessage")
+        public List<String> duplicatedMessage(@Source List<TestPojo> pojos) {
+            if (!this.injectedBean.getId().equals(this.injectedBeanId)) {
+                throw new IllegalStateException("duplicatedMessage must be executed in the same request context as getPojos");
+            }
+            return pojos.stream()
+                    .map(pojo -> pojo.getMessage() + pojo.getMessage())
+                    .collect(Collectors.toList());
+        }
+
+        /**
+         * This source field duplicates the message of a TestPojo (repeats it twice) and does it asynchronously.
+         */
+        @Name("duplicatedMessageAsync")
+        public CompletionStage<List<String>> duplicatedMessageAsync(@Source List<TestPojo> pojos) {
+            if (!this.injectedBean.getId().equals(this.injectedBeanId)) {
+                throw new IllegalStateException(
+                        "duplicatedMessageAsync must be executed in the same request context as getPojos");
+            }
+            return CompletableFuture.completedFuture(pojos.stream()
+                    .map(pojo -> pojo.getMessage() + pojo.getMessage())
+                    .collect(Collectors.toList()));
+        }
+
+    }
+
+    @RequestScoped
+    public static class InjectedBean {
+
+        private Long id = ThreadLocalRandom.current().nextLong();
+
+        public Long getId() {
+            return id;
+        }
+    }
+
+}

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/SecurityTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/SecurityTest.java
@@ -4,11 +4,16 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
 import javax.annotation.security.RolesAllowed;
 import javax.enterprise.context.ApplicationScoped;
 
 import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Name;
 import org.eclipse.microprofile.graphql.Query;
+import org.eclipse.microprofile.graphql.Source;
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.asset.EmptyAsset;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
@@ -24,7 +29,7 @@ public class SecurityTest extends AbstractGraphQLTest {
     @RegisterExtension
     static QuarkusUnitTest test = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClass(SecuredApi.class)
+                    .addClasses(SecuredApi.class, Foo.class)
                     .addAsResource("application-secured.properties", "application.properties")
                     .addAsResource("users.properties")
                     .addAsResource("roles.properties")
@@ -32,7 +37,7 @@ public class SecurityTest extends AbstractGraphQLTest {
 
     @Test
     public void testAuthenticatedUser() {
-        String query = getPayload("{ foo }");
+        String query = getPayload("{ foo { message} }");
         RestAssured.given()
                 .header(new Header("Authorization", "Basic ZGF2aWQ6cXdlcnR5MTIz"))
                 .body(query)
@@ -41,12 +46,26 @@ public class SecurityTest extends AbstractGraphQLTest {
                 .then()
                 .assertThat()
                 .body("errors", nullValue())
-                .body("data.foo", equalTo("foo"));
+                .body("data.foo.message", equalTo("foo"));
+    }
+
+    @Test
+    public void testAuthenticatedUserWithSource() {
+        String query = getPayload("{ foo { bonusFoo } }");
+        RestAssured.given()
+                .header(new Header("Authorization", "Basic ZGF2aWQ6cXdlcnR5MTIz"))
+                .body(query)
+                .contentType(MEDIATYPE_JSON)
+                .post("/graphql/")
+                .then()
+                .assertThat()
+                .body("errors", nullValue())
+                .body("data.foo.bonusFoo", equalTo("bonus"));
     }
 
     @Test
     public void testUnauthorizedRole() {
-        String query = getPayload("{ bar }");
+        String query = getPayload("{ bar { message } }");
         RestAssured.given()
                 .header(new Header("Authorization", "Basic ZGF2aWQ6cXdlcnR5MTIz"))
                 .body(query)
@@ -55,7 +74,43 @@ public class SecurityTest extends AbstractGraphQLTest {
                 .then()
                 .assertThat()
                 .body("errors", notNullValue())
-                .body("data.bar", nullValue());
+                .body("data.bar.message", nullValue());
+    }
+
+    /**
+     * Call a query that we are authorized to call, but within that, retrieve a source field that we're not authorized
+     * to retrieve.
+     */
+    @Test
+    public void testUnauthorizedForSource() {
+        String query = getPayload("{ foo { bonusBar } }");
+        RestAssured.given()
+                .header(new Header("Authorization", "Basic ZGF2aWQ6cXdlcnR5MTIz"))
+                .body(query)
+                .contentType(MEDIATYPE_JSON)
+                .post("/graphql/")
+                .then()
+                .assertThat()
+                .body("errors", notNullValue())
+                .body("data.foo.bonusBar", nullValue());
+    }
+
+    static class Foo {
+
+        private String message;
+
+        public Foo(String foo) {
+            this.message = foo;
+        }
+
+        public String getMessage() {
+            return message;
+        }
+
+        public void setMessage(String message) {
+            this.message = message;
+        }
+
     }
 
     @GraphQLApi
@@ -64,14 +119,26 @@ public class SecurityTest extends AbstractGraphQLTest {
 
         @Query
         @RolesAllowed("fooRole")
-        public String foo() {
-            return "foo";
+        public Foo foo() {
+            return new Foo("foo");
+        }
+
+        @Name("bonusFoo")
+        @RolesAllowed("fooRole")
+        public List<String> bonusFoo(@Source List<Foo> foos) {
+            return foos.stream().map(foo -> "bonus").collect(Collectors.toList());
         }
 
         @Query
         @RolesAllowed("barRole")
-        public String bar() {
-            return "bar";
+        public Foo bar() {
+            return new Foo("bar");
+        }
+
+        @Name("bonusBar")
+        @RolesAllowed("barRole")
+        public List<String> bonusBar(@Source List<Foo> foos) {
+            return foos.stream().map(foo -> "bonus").collect(Collectors.toList());
         }
 
     }

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLContextTestCase.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLContextTestCase.java
@@ -1,0 +1,256 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.context.ManagedExecutor;
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Name;
+import org.eclipse.microprofile.graphql.Query;
+import org.eclipse.microprofile.graphql.Source;
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
+import io.smallrye.graphql.api.Context;
+
+public class SmallRyeGraphQLContextTestCase extends AbstractGraphQLTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(Dummy.class, ContextPropagationResource.class));
+
+    @Test
+    public void testAsyncQuery() {
+        String request = getPayload("{\n" +
+                "  testAsyncQuery {\n" +
+                "    stringField\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(request)
+                .post("/graphql")
+                .then()
+                .log()
+                .body(true)
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.testAsyncQuery.stringField", Matchers.equalTo("OK"));
+    }
+
+    @Test
+    public void testAsyncQueryWithAsyncSource() {
+        String request = getPayload("{\n" +
+                "  testAsyncQuery {\n" +
+                "    stringField\n" +
+                "    stringBatchSourceAsync\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(request)
+                .post("/graphql")
+                .then()
+                .log()
+                .body(true)
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.testAsyncQuery.stringField", Matchers.equalTo("OK"))
+                .body("data.testAsyncQuery.stringBatchSourceAsync", Matchers.equalTo("hello"));
+    }
+
+    @Test
+    public void testManualPropagation() {
+        String request = getPayload("{\n" +
+                "  testManualPropagation {\n" +
+                "    stringField\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(request)
+                .post("/graphql")
+                .then()
+                .log()
+                .body(true)
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.testManualPropagation.stringField", Matchers.equalTo("OK"));
+    }
+
+    @Test
+    public void testSourceMethods() {
+        String request = getPayload("{\n" +
+                "  testSourceMethods {\n" +
+                "    stringBatchSource,\n" +
+                "    numberBatchSource\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body(request)
+                .post("/graphql")
+                .then()
+                .log()
+                .body(true)
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body("data.testSourceMethods.stringBatchSource", Matchers.equalTo("hello"));
+    }
+
+    @GraphQLApi
+    @ApplicationScoped
+    public static class ContextPropagationResource {
+
+        @Inject
+        Instance<Context> context;
+
+        @Inject
+        ManagedExecutor managedExecutor;
+
+        @Query
+        public CompletableFuture<Dummy> testAsyncQuery() {
+            final String executionId = context.get().getExecutionId();
+            assertNotNull(executionId);
+            this.executionId = context.get().getExecutionId();
+            assertEquals("testAsyncQuery", context.get().getFieldName());
+            Dummy result = new Dummy();
+            result.setStringField("OK");
+            return CompletableFuture.completedFuture(result);
+        }
+
+        @Query
+        public Dummy testManualPropagation() throws ExecutionException, InterruptedException {
+            final String executionId = context.get().getExecutionId();
+            assertNotNull(executionId);
+
+            CompletableFuture<Integer> numberFuture = CompletableFuture.supplyAsync(() -> {
+                assertNotNull(context);
+                assertEquals(executionId, context.get().getExecutionId());
+                assertEquals("testManualPropagation", context.get().getFieldName());
+                return 42;
+            }, managedExecutor);
+
+            CompletableFuture<String> stringFuture = CompletableFuture.supplyAsync(() -> {
+                assertNotNull(context, "Context must to be available inside an async task");
+                assertEquals(executionId, context.get().getExecutionId(), "Execution ID must be the same inside an async task");
+                assertEquals("testManualPropagation", context.get().getFieldName());
+                return "OK";
+            }, managedExecutor);
+
+            Dummy result = new Dummy();
+            result.setNumberField(numberFuture.get());
+            result.setStringField(stringFuture.get());
+            return result;
+        }
+
+        private volatile String executionId;
+
+        @Query
+        public Dummy testSourceMethods() {
+            this.executionId = context.get().getExecutionId();
+            assertEquals("testSourceMethods", context.get().getFieldName());
+            return new Dummy();
+        }
+
+        @Name("stringBatchSource")
+        public List<String> stringBatchSource(@Source List<Dummy> source) {
+            assertEquals("stringBatchSource", context.get().getFieldName());
+            assertEquals(this.executionId, context.get().getExecutionId(), "Wrong execution ID propagated from the root query");
+
+            List<String> result = new ArrayList<>();
+            for (Dummy dummy : source) {
+                result.add("hello");
+            }
+            return result;
+        }
+
+        @Name("stringBatchSourceAsync")
+        public CompletionStage<List<String>> stringBatchSourceAsync(@Source List<Dummy> source) {
+            assertEquals("stringBatchSourceAsync", context.get().getFieldName());
+            assertEquals(this.executionId, context.get().getExecutionId(), "Wrong execution ID propagated from the root query");
+
+            List<String> result = new ArrayList<>();
+            for (Dummy dummy : source) {
+                result.add("hello");
+            }
+            return CompletableFuture.completedFuture(result);
+        }
+
+        @Name("numberBatchSource")
+        public List<Integer> numberBatchSource(@Source List<Dummy> source) {
+            assertEquals("numberBatchSource", context.get().getFieldName());
+            assertEquals(this.executionId, context.get().getExecutionId(), "Wrong execution ID propagated from the root query");
+
+            List<Integer> result = new ArrayList<>();
+            for (Dummy dummy : source) {
+                result.add(123);
+            }
+            return result;
+        }
+
+    }
+
+    public static class Dummy {
+
+        private String stringField;
+
+        private Integer numberField;
+
+        private Boolean booleanField;
+
+        public String getStringField() {
+            return stringField;
+        }
+
+        public void setStringField(String stringField) {
+            this.stringField = stringField;
+        }
+
+        public Integer getNumberField() {
+            return numberField;
+        }
+
+        public void setNumberField(Integer numberField) {
+            this.numberField = numberField;
+        }
+
+        public Boolean getBooleanField() {
+            return booleanField;
+        }
+
+        public void setBooleanField(Boolean booleanField) {
+            this.booleanField = booleanField;
+        }
+    }
+
+}

--- a/extensions/smallrye-graphql/runtime/pom.xml
+++ b/extensions/smallrye-graphql/runtime/pom.xml
@@ -35,6 +35,10 @@
             <groupId>io.smallrye</groupId>
             <artifactId>smallrye-graphql-cdi</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-context-propagation</artifactId>
+        </dependency>
     </dependencies>
 
     <build>

